functions/src/handleFormSubmit.ts

```diff
@@ -1,36 +1,118 @@
-exports.handler = async function(request: any, response: any, database: any) {
-  console.log(request.body);
-
-  const name = request.body['name'];
-  const describeYourself = request.body['describeYourself'];
-  const email = request.body['email'];
-  const batch = request.body['batch'];
-  const scholarship = request.body['scholarship'];
-  const consent = request.body['consent'];
-  const track = request.body['track'];
-  // TODO change to proper name
-  const learnAboutSource = request.body['application-form'];
-
-  try {
-    await database
-      .firestore()
-      .collection('batches')
-      .doc('batch-'.concat(batch))
-      .collection('applications')
-      .add({
-        name: name,
-        email: email,
-        userType: describeYourself,
-        batch: batch,
-        scholarship: scholarship,
-        consent: consent,
-        track: track,
-        source: learnAboutSource,
-      });
-
-    response.send('Thank you for your application');
-  } catch (error) {
-    console.log(error);
-    response.status(500).send(error);
-  }
+const Busboy = require('busboy');
+const cors = require('cors');
+const fs = require('fs');
+const STORAGE_BUCKET_URL = require('./constants').STORAGE_BUCKET_URL;
+
+// TODO This should be in own file, because it would be used accross
+// all functions
+const whitelist = [
+  'http://localhost:8000',
+  'https://dp.school',
+  'https://dpschool.io',
+  'http://digitalproductschool.io',
+];
+const corsOptions = {
+  origin: (origin: string, callback: any) => {
+    if (whitelist.includes(origin)) {
+      callback(null, true);
+    } else {
+      callback(new Error('Not allowed by CORS'));
+    }
+  },
+};
+
+const corsFn = cors(corsOptions);
+
+// busboy is a middleware to collect multipart/form-data
+// it is finished once all file readstreams have been processed
+// we have two options: either upload to cloud while reading the stream
+// or read the stream and return it as binary to other function that
+// could save it to the cloud. Currently, we read the stream & directly
+// save / pipe to the cloud storage. This means the functions have two
+// responsibilities: getting all key: value pairs and at the same time
+// piping files to the storage.
+exports.getRequestDataAndPersistFiles = async (
+  req: any,
+  database: any
+): Promise<any> => {
+  return new Promise((resolve, reject) => {
+    const busboy = new Busboy({ headers: req.headers });
+    const data: { [key: string]: string | Buffer | Promise<any> } = {};
+
+    // collect file stream and pipe them to the storage, return promises as values
+    busboy.on(
+      'file',
+      async (fieldname: string, readFileStream: any, filename: string) => {
+        const filePromise = new Promise(async (fileResolve, fileReject) => {
+          const bucket = STORAGE_BUCKET_URL;
+          const name = `participants/${Date.now()}_${filename}`;
```

Comment:
I'd like to have a similar path like we have in the database. 
So first a folder for each batch and one folder for each participant. In this one, we can store the CV and/or cover letter. 

This way, we can track more easily whether the CVs get deleted after the grace period.

Reply:
yep!